samza-core/src/main/java/org/apache/samza/system/inmemory/InMemorySystemFactory.java

```diff
@@ -31,20 +33,28 @@
  * Initial draft of in-memory {@link SystemFactory}. It is test only and not meant for production use right now.
  */
 public class InMemorySystemFactory implements SystemFactory {
-  private static final InMemoryManager MEMORY_MANAGER = new InMemoryManager();
+  private static final ConcurrentHashMap<String, InMemoryManager> IN_MEMORY_MANAGERS = new ConcurrentHashMap<>();
 
   @Override
   public SystemConsumer getConsumer(String systemName, Config config, MetricsRegistry registry) {
-    return new InMemorySystemConsumer(MEMORY_MANAGER);
+    return new InMemorySystemConsumer(getOrDefaultInMemoryManagerByTestId(config));
   }
 
   @Override
   public SystemProducer getProducer(String systemName, Config config, MetricsRegistry registry) {
-    return new InMemorySystemProducer(systemName, MEMORY_MANAGER);
+    return new InMemorySystemProducer(systemName, getOrDefaultInMemoryManagerByTestId(config));
   }
 
   @Override
   public SystemAdmin getAdmin(String systemName, Config config) {
-    return new InMemorySystemAdmin(MEMORY_MANAGER);
+    return new InMemorySystemAdmin(getOrDefaultInMemoryManagerByTestId(config));
+  }
+
+  private InMemoryManager getOrDefaultInMemoryManagerByTestId(Config config) {
+    String inMemoryScope = config.get(InMemorySystemConfig.INMEMORY_SCOPE);
+    if (inMemoryScope == null) {
+      inMemoryScope = InMemorySystemConfig.DEFAULT_INMEMORY_SCOPE;
+    }
```

Comment:
We can move this logic to `InMemorySystemConfig` where if the config is present we return the value if not we return the default scope. By doing this, the SystemFactory will just need to get the scope and use that directly for fetch the value from map.

This will offer consistency in terms of what we default to in the absence of scope plus it eliminates the duplication of defaulting logic in multiple places.

Reply:
Agreed!